consensus/types/peer_round_state.go

```diff
@@ -46,6 +46,29 @@ func (prs PeerRoundState) String() string {
 	return prs.StringIndented("")
 }
 
+// Copy provides a deep copy operation. Because many of the fields in
+// the PeerRound struct are pointers, we need an explicit deep copy
+// operation to avoid a non-obvious shared data situation.
+func (prs PeerRoundState) Copy() PeerRoundState {
+	// this works because it's not a pointer receiver so it's
+	// already, effectively a copy.
+
```

Comment:
nit: we can probably store `prs.ProposalBlockPartSetHeader.Hash.Bytes()` in a variable and use that for length and copying instead of calling the API twice.